### PR TITLE
#27: Improved quoting and cleaned up handling of placeholders and parameters

### LINKED ISSUE #19: Add a method for quoting with double quotes
## Problem

We use double quotes a lot in virtual schemas for identifiers (table names, schema names etc). We need to discuss in the team if we want to switch to a single quoting for identifiers too or we can add a method that applies double quotes to a builder.

### LINKED ISSUE #27: Add forced quoting option
## Situation

We have automatic quoting and optional unquoted replacement. Generally for unknown object types not using quotes is fine. But there are exceptions. So forcing quotes with ``{{placeholder|q}}` would be useful.

## Acceptance Criteria

1. Forced quoting with `q` switch supported.

### PR DESCRIPTION
**Pull request by redcatbear**
Closes #27.
Closes #19.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -44,37 +44,58 @@ Result:
...
+    E-TEST-2: Message with 'q-value' and uq-value.
+
+#### Automatic Quoting
+
+When replacing placeholders in messages, `ExaError` quotes the values according to your choices. If you don't specify a quoting option in a placeholder, you get auto-quoting. In this mode values are quoted depending on their type.
+
+| Type            | Quoted with     | Example                    | Since version |
+|-----------------|-----------------|----------------------------|--------------:|
+| `String`        | single quotes   | `'Hello world!'`           |               |
+| `Character`     | single quotes   | `'A'`                      |               |
```
Comment (kaklakariada) [inline README.md]: also for `char`?

File: README.md [hunk 2]
```diff
@@ -44,37 +44,58 @@ Result:
...
 ```
 
 Result:
 
-    E-TEST-2: Unknown input: unknown.
+    E-TEST-2: Message with 'q-value' and uq-value.
+
+#### Automatic Quoting
+
+When replacing placeholders in messages, `ExaError` quotes the values according to your choices. If you don't specify a quoting option in a placeholder, you get auto-quoting. In this mode values are quoted depending on their type.
```
Comment (kaklakariada) [inline README.md:60]: 👍

File: src/main/java/com/exasol/errorreporting/ErrorMessageBuilder.java [hunk 3]
```diff
@@ -98,14 +67,19 @@ public ErrorMessageBuilder parameter(final String placeholder, final Object valu
      * <p>
      * You can use the parameter in message and mitigation using {@code {{parameter}}}.
      * </p>
+     * <p>
+     * Note that the last parameter exists only as a means to add a description in the error catalog. It is not used
+     * when displaying error messages to the application users.
```
Comment (kaklakariada) [inline src/main/java/com/exasol/errorreporting/ErrorMessageBuilder.java:72]: 👍

File: src/main/java/com/exasol/errorreporting/Placeholder.java [hunk 4]
```diff
@@ -4,62 +4,152 @@
...
+    private final String reference;
+    private final int startIndex;
+    private final int endIndex;
+    private final Quoting quoting;
 
     /**
-     * Create a new instance of {@link Placeholder}.
+     * Parse a placeholder from a placeholder string.
+     * <p>
+     * Placeholders a reference part and a part that contains switched that control how the placeholder is treated. This
```
Comment (kaklakariada) [inline src/main/java/com/exasol/errorreporting/Placeholder.java]:
```suggestion
     * Placeholders consist of a reference part and a part that contains switches that control how the placeholder is treated. This
```
❓

File: src/main/java/com/exasol/errorreporting/Quoter.java [hunk 5]
```diff
@@ -14,21 +14,36 @@ private Quoter() {
...
      * @return quoted object
      */
-    static String quoteObject(final Object object) {
+    static String quoteObject(final Object object, final Quoting quoting) {
         if (object == null) {
             return "<null>";
-        } else if (object instanceof String || object instanceof Character || object instanceof java.nio.file.Path
-                || object instanceof java.io.File || object instanceof java.net.URL || object instanceof java.net.URI) {
-            return "'" + object + "'";
         } else if (object instanceof List) {
```
Comment (kaklakariada) [inline src/main/java/com/exasol/errorreporting/Quoter.java]: You could also use `Collection` or `Iterable` here, then we can handle all types of collections.

File: src/main/java/com/exasol/errorreporting/Quoter.java [hunk 6]
```diff
@@ -14,21 +14,36 @@ private Quoter() {
...
             final List<?> list = (List<?>) object;
-            return "[" + list.stream().map(Quoter::quoteObject).collect(Collectors.joining(", ")) + "]";
+            return "[" + list.stream().map(object1 -> quoteObject(object1, quoting))
```
Comment (kaklakariada) [inline src/main/java/com/exasol/errorreporting/Quoter.java]:
```suggestion
            return "[" + list.stream().map(item -> quoteObject(item, quoting))
```
❓

File: src/test/java/com/exasol/errorreporting/ParameterDefinitionListTest.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.sameInstance;
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+class ParameterDefinitionListTest {
+    @Test
+    void testGet() {
```
Comment (kaklakariada) [inline src/test/java/com/exasol/errorreporting/ParameterDefinitionListTest.java:12]: Add a test that adds two parameter definitions with the same name, assert that only the first one is returned.

File: src/test/java/com/exasol/errorreporting/PlaceholderTest.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+class PlaceholderTest {
+    @Test
+    void TestParseParameterNameWithoutSwitches()
+    {
+        final Placeholder placeholder = Placeholder.parse("the_parameter").build();
+        assertAll(
+                () -> assertThat(placeholder.getReference(), equalTo("the_parameter")),
+                () -> assertThat(placeholder.getQuoting(), equalTo(Quoting.AUTOMATIC))
```
Comment (kaklakariada) [inline src/test/java/com/exasol/errorreporting/PlaceholderTest.java]: This is duplicated in all tests. Maybe extract this into a method `assertParameter()`?

File: src/test/java/com/exasol/errorreporting/QuoterTest.java [hunk 9]
```diff
@@ -20,14 +28,87 @@ static Stream<Arguments> getQuotingExamples() {
...
+        return Stream.of(//
+                Arguments.of(1, "1"), //
+                Arguments.of(1L, "1"), //
+                Arguments.of(1f, "1.0"), //
+                Arguments.of(1d, "1.0"), //
+                Arguments.of(List.of(1, "test"), "[1, test]"), //
+                Arguments.of(new CustomObject(), "<CustomObject>"), //
+                Arguments.of(null, "<null>"), //
+                Arguments.of("test", "test"), //
+                Arguments.of(new File(SEP + "foo" + SEP + "baz.txt"), SEP + "foo" + SEP + "baz.txt"), //
```
Comment (kaklakariada) [inline src/test/java/com/exasol/errorreporting/QuoterTest.java]: This test is platform dependent.
```suggestion
                Arguments.of(new File("/foo/baz.txt"), "/foo/baz.txt"), //
                Arguments.of(new File("\\foo\\baz.txt"), "\\foo\\baz.txt"), //
```
(same for the following tests)

File: src/test/java/com/exasol/errorreporting/QuoterTest.java [hunk 10]
```diff
@@ -20,14 +28,87 @@ static Stream<Arguments> getQuotingExamples() {
                 Arguments.of(List.of(1, "test"), "[1, 'test']"), //
                 Arguments.of(new CustomObject(), "<CustomObject>"), //
                 Arguments.of(null, "<null>"), //
-                Arguments.of("test", "'test'")//
+                Arguments.of("test", "'test'"), //
+                Arguments.of(new File(SEP + "foo" + SEP + "baz.txt"), "'"+ SEP + "foo" + SEP + "baz.txt'"), //
```
Comment (kaklakariada) [inline src/test/java/com/exasol/errorreporting/QuoterTest.java]: This test is platform dependent.
```suggestion
                Arguments.of(new File("/foo/baz.txt"), "'/foo/baz.txt'"), //
                Arguments.of(new File("\\foo\\baz.txt"), "'\\foo\\baz.txt'"), //
```

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (redcatbear) [reply README.md]: Yes. Added test and documentation.

File: src/main/java/com/exasol/errorreporting/Quoter.java (hunk 5 above)
Comment (redcatbear) [reply src/main/java/com/exasol/errorreporting/Quoter.java]: Good idea. Done.

File: src/test/java/com/exasol/errorreporting/ParameterDefinitionListTest.java (hunk 7 above)
Comment (redcatbear) [reply src/test/java/com/exasol/errorreporting/ParameterDefinitionListTest.java:12]: Done.

File: src/test/java/com/exasol/errorreporting/PlaceholderTest.java (hunk 8 above)
Comment (redcatbear) [reply src/test/java/com/exasol/errorreporting/PlaceholderTest.java]: Done.

**Review by kaklakariada — APPROVED** [review]

**Pull request merged**